### PR TITLE
[Sofa.Type] Clean and modernize Vec and Mat

### PR DESCRIPTION
**Pull request by fredroy**
- Using noexcept and constexpr keywords here and there.
- replace multiple versions of functions/ctors with conditional templated ones 
- cleaning (static_cast, etc)

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Mat.h [hunk 1]
```diff
@@ -114,22 +97,22 @@ class Mat : public fixed_array<VecNoInit<C,real>, L>
...
 
     /// Constructor from another matrix with different size (with null default entries and ignoring outside entries)
     template<Size L2, Size C2, typename real2>
-    explicit Mat(const Mat<L2,C2,real2>& m)
+    explicit constexpr Mat(const Mat<L2,C2,real2>& m) noexcept
     {
-        Size maxL = std::min( L, L2 );
-        Size maxC = std::min( C, C2 );
+        constexpr Size maxL = std::min( L, L2 );
+        constexpr Size maxC = std::min( C, C2 );
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Mat.h]:
```suggestion
        constexpr Size minL = std::min( L, L2 );
        constexpr Size minC = std::min( C, C2 );
```
Funny that they are called `max` whereas they compute the min

**Review by alxbilger — CHANGES_REQUESTED** [review]
Almost good to go. Congrats

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Mat.h (hunk 1 above)
Comment (fredroy) [reply SofaKernel/modules/Sofa.Type/src/sofa/type/Mat.h]: cannot commit your suggestion, as I have to change the subsequent loops (using those variables) :/

**Pull request merged**